sdk/nodejs/cmd/run/run.ts

```diff
@@ -344,26 +390,31 @@ ${defaultMessage}`,
         const runProgramSpan = tracing.newSpan("language-runtime.runProgram");
 
         try {
-            const projectRoot = projectRootFromProgramPath(program);
-            const packageObject = packageObjectFromProjectRoot(projectRoot);
-
+            const packageRoot = await npmPackageRootFromProgramPath(program);
+            const packageObject = packageObjectFromProjectRoot(packageRoot);
             let programExport: any;
 
+            // If the user provided an entrypoint, we use that file
+            // relative to the package directory.
+            if (packageObject["main"]) {
+                const packageMainPath = path.join(packageRoot, packageObject["main"]);
+                if (fs.existsSync(packageMainPath)) {
+                    program = packageMainPath;
+                } else {
+                    log.warn(
+                        `Could not find entry point '${packageMainPath}' specified in package.json; ` +
+                            `using '${program}' instead`,
+                    );
+                }
```

Comment:
This is the difference from the previous attempt.

The [previous attempt](https://github.com/pulumi/pulumi/pull/12759/files#diff-56150bcb0a988cc1b28983a197010bbca7a34e91910a23c5062aee074e8164b6R398) always set `program = path.join(packageRoot, packageObject["main"])`, which would not work if the path specified in `package.json`'s `main` does not exist, which was the case for some of our older tests that specify a nonexistent `bin/index.js` for `main`.

Now, we check to see if what is specified in `main` exists. If it exists, we use it. Otherwise, we emit a warning and continue with the previous behavior.